### PR TITLE
Update JSON Patch Functionality

### PR DESCRIPTION
**Pull request by alzimmermsft**
This PR updates JSON Patch functionality in `azure-core-experimental`.

- Added support for `JsonPatchDocument` to accept a `JsonSerializer` during construction of the class.
- Changed `value` parameters from JSONified strings to Object.
- Changed handling of `value` parameters to allow for a JSON null to be serialized as the value to support all scenarios in JSON Patch.
- Added `JsonPatchDocumentSerializer` and `JsonPatchOperationSerializer` to handle serialization of `JsonPatchDocument` and `JsonPatchOperation`.
- Made `JsonPatchOperation` and `JsonPatchOperationKind` public.
- Added `getJsonPatchOperations` to `JsonPatchDocument` to get a read only list of operations in the document.

### CODE REVIEW

**Review by srnagar — DISMISSED** [review]
LGTM!

File: sdk/core/azure-core-experimental/src/main/java/com/azure/core/experimental/jsonpatch/JsonPatchDocument.java [hunk 1]
```diff
@@ -22,15 +25,44 @@
...
         this.operations = new ArrayList<>();
+        this.serializer = serializer;
+    }
+
+    /**
+     * Gets the JSON Patch operations in this document.
+     *
+     * @return The JSON Patch operations in this document.
+     */
+    public List<Object> getJsonPatchOperations() {
```
Comment (srnagar) [inline sdk/core/azure-core-experimental/src/main/java/com/azure/core/experimental/jsonpatch/JsonPatchDocument.java:64]: Can this return a `List<JsonPatchOperation>` instead?

**Review by srnagar — APPROVED** [review]

File: sdk/core/azure-core-experimental/src/main/java/com/azure/core/experimental/jsonpatch/JsonPatchDocument.java (hunk 1 above)
Comment (alzimmermsft) [reply sdk/core/azure-core-experimental/src/main/java/com/azure/core/experimental/jsonpatch/JsonPatchDocument.java]: Done

**Review by srnagar — APPROVED** [review]

File: sdk/core/azure-core-experimental/src/main/java/com/azure/core/experimental/jsonpatch/JsonPatchDocument.java [hunk 2]
```diff
@@ -65,19 +87,17 @@ public JsonPatchDocument appendAdd(String path, String rawJsonValue) {
...
      *
      * @param path The path to replace.
-     * @param rawJsonValue The raw JSON value to use as the replacement.
+     * @param value The value to use as the replacement.
      * @return The updated JsonPatchDocument object.
-     * @throws NullPointerException If {@code path} or {@code rawJsonValue} is null.
+     * @throws NullPointerException If {@code path} is null.
      */
-    public JsonPatchDocument appendReplace(String path, String rawJsonValue) {
+    public JsonPatchDocument appendReplace(String path, Object value) {
```
Comment (pakrym) [inline sdk/core/azure-core-experimental/src/main/java/com/azure/core/experimental/jsonpatch/JsonPatchDocument.java:97]: I think switching the behavior of this parameter based on ctor parameter might be a little confusing.

Imagine you were replacing some trivial properties in the beginning but now wanting to pass a serialized object into another property. Passing the serializer into ctor would change the behavior of all existing code.

**Comment by pakrym** [comment]
AFAIK You have a bunch of JSON DOM interfaces in core, do we want to take those as well?

File: sdk/core/azure-core-experimental/src/main/java/com/azure/core/experimental/jsonpatch/JsonPatchDocument.java (hunk 2 above)
Comment (alzimmermsft) [reply sdk/core/azure-core-experimental/src/main/java/com/azure/core/experimental/jsonpatch/JsonPatchDocument.java:97]: Do we foresee it being a common scenario for multiple seriliazers to be used when interacting with a single `JsonPatchDocument`? With the pattern being the patch document is tied to a single serializer I would think that a different document would be used.

Another possible solution to this problem, given `JsonPatchOperation` is in public API is to add an `appendOperation(JsonPatchOperation operation)` API where the customer would set the `optionalValue` manually if they wanted to use multiple serializers for a single document.

**Comment by alzimmermsft** [comment]
> 
> 
> AFAIK You have a bunch of JSON DOM interfaces in core, do we want to take those as well?

Is this in reference to `JsonObject` and the other interfaces? If so, I won't be eagerly consuming those as we haven't had a great use case for moving those into `azure-core` any time soon.

File: sdk/core/azure-core-experimental/src/main/java/com/azure/core/experimental/jsonpatch/JsonPatchDocument.java (hunk 2 above)
Comment (pakrym) [reply sdk/core/azure-core-experimental/src/main/java/com/azure/core/experimental/jsonpatch/JsonPatchDocument.java:97]: I was confused as to what default behavior is.

**Pull request merged**